### PR TITLE
End-to-end "pp-spartan" verification contract (Step 4, Step 5)

### PR DESCRIPTION
**Pull request by storojs72**
This PR introduces two additional libraries for Step 4 and Step 5 of end-to-end verification contract. Code for Step 4 is pretty straightforward, so it is actually evaluated via passing CI integration test. Step 5 is more complicated, so there is a set of unit tests specifically for it. Also, Step 5 requires IdentityPolynomial building block, which is also included.

Note for reviewers: as usually, please pay closest attention on correspondence of the Solidity implementation to the Rust [reference](https://github.com/lurk-lab/Nova/blob/solidity-verifier-pp-spartan/src/spartan/ppsnark.rs#L1565) one.

### CODE REVIEW

File: src/verifier/Step5.sol [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+        }
+
+        uint8[] memory label = new uint8[](1);
+        label[0] = 0x65; // Rust's b"e"
+
+        transcript = KeccakTranscriptLib.absorb(transcript, label, evals);
+
+        label[0] = 0x72; // Rust's b"r"
+
+        // num_claims is hardcoded to be equal to 10
```
Comment (huitseeker) [inline src/verifier/Step5.sol:273]: is there a way to relax this requirement?

File: test/pp-spartan-step5-tests.t.sol [hunk 2]
```diff
@@ -0,0 +1,998 @@
...
+        evals[15] = 0x01aba07856c3668fbdaa165d70fa7539e6ca2ae116e959fca79f879c70259fa6;
+
+        uint8[] memory label = new uint8[](1);
+        label[0] = 0x65; // Rust's b"e"
+
+        transcript = KeccakTranscriptLib.absorb(transcript, label, evals);
+
+        label[0] = 0x72; // Rust's b"r"
+
+        // num_claims is hardcoded to be equal to 10
```
Comment (huitseeker) [inline test/pp-spartan-step5-tests.t.sol:973]: In tests, that's fine!

File: src/verifier/Step5.sol [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
```
Comment (huitseeker) [inline src/verifier/Step5.sol:237]: Could we relax this?

File: src/verifier/Step5.sol (hunk 1 above)
Comment (storojs72) [reply src/verifier/Step5.sol:273]: Not sure if I follow. This is hardcoded in the [reference](https://github.com/lurk-lab/Nova/blob/solidity-verifier-pp-spartan/src/spartan/ppsnark.rs#L1651) implementation as well.

File: src/blocks/IdentityPolynomial.sol [hunk 4]
```diff
@@ -0,0 +1,19 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.16;
+
+import "src/blocks/pasta/Vesta.sol";
+import "src/blocks/pasta/Pallas.sol";
+
+// Port of Nova' SparsePolynomial (https://github.com/microsoft/Nova/blob/main/src/spartan/polynomial.rs#L121)
```
[the remark this answers is not in the document]
Comment (storojs72) [reply src/blocks/IdentityPolynomial.sol:7]: Oh.. agree, the comment is wrong

File: src/verifier/Step5.sol (hunk 1 above)
Comment (huitseeker) [reply src/verifier/Step5.sol]: Oh, yes, I'm a dunce, and yes, the link to the reference Rust would definitely have helped silly me figure it out!

**Review by huitseeker — APPROVED** [review]
Thank you very much, this LGTM!

**Pull request merged**